testutil/network/config.go

```diff
@@ -0,0 +1,86 @@
+package network
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	db "github.com/cometbft/cometbft-db"
+	"github.com/cometbft/cometbft/libs/rand"
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	"github.com/cosmos/cosmos-sdk/crypto/hd"
+	"github.com/cosmos/cosmos-sdk/crypto/keyring"
+	sdkservertypes "github.com/cosmos/cosmos-sdk/server/types"
+	pruninttypes "github.com/cosmos/cosmos-sdk/store/pruning/types"
+	"github.com/cosmos/cosmos-sdk/testutil/network"
+	"github.com/cosmos/cosmos-sdk/testutil/sims"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/types"
+
+	"github.com/pokt-network/poktroll/app"
+)
+
+// GetNumApplications returns the number of applications to be created in the network at genesis.
+func (cfg *InMemoryNetworkConfig) GetNumApplications(t *testing.T) int {
+	t.Helper()
+
+	if cfg.NumApplications > 0 {
+		return cfg.NumApplications
+	}
+
+	return cfg.AppSupplierPairingRatio * cfg.NumSuppliers
+}
+
+// GetNumKeyringAccounts returns the number of keyring accounts needed for the given configuration.
+func (cfg *InMemoryNetworkConfig) GetNumKeyringAccounts(t *testing.T) int {
+	t.Helper()
+
+	return cfg.NumGateways + cfg.NumSuppliers + cfg.GetNumApplications(t)
```

Comment:
This has to do with how the different network implementations use the shared network config. In the case of `inMemoryNetworkWithSessions`, `#GetNumApplications()` MUST be used because it is invalid to provide `NumApplications` with it (use `AppToSupplierRatio` instead). This is enforced in the constructor; `inMemoryNetworkWithGateways` has the opposite constraint. My goal was to minimize complexity by sharing a single config with minimal necessary fields that are common to all network implementations.

Reply:
If `NumApplications` is not populated in the config when it is provided, it is never subsequently populated by network helpers. My intention was for `NumApplications` and `AppToSupplierRatio` to be inputs and `#GetNumApplications() to be the output (single source of truth). I'm reluctant to have the helpers mutate the config after it's provided. That just seems like something that would come back around to bite us later.

I  will add a comment saying as much, let me know what you think.If `NumApplications` is not populated in the config when it's provided, it is never subsequently populated by network helpers. My intention was for `NumApplications` and `AppToSupplierRatio` to be inputs and `#GetNumApplications()` to be the output (source of truth). I'm reluctant to have helpers mistress the config after it's provided to the network constructor. They just send like something that would come back around to bite us IMO.